qpmodel/optimizer.cs

```diff
@@ -349,95 +431,188 @@ public void PropagateProperty(PhysicProperty property)
         }
 
         // scan through the member list and return the least cost member
-        public CGroupMember locateMinCostMember()
+        public CGroupMember locateMinCostMember(PhysicProperty required)
         {
-            double mincost = double.MaxValue;
+            double mincost = 0, propmincost = 0;
+            if (required != null) propmincost = double.MaxValue;
+            // calculate null requirement min cost if not available
+            if (minMember_.Count == 0) mincost = double.MaxValue;
+
             foreach (var v in exprList_)
             {
                 var physic = v.physic_;
-                if (physic != null && physic.Cost() < mincost)
+                if (physic is null) continue;
+                // make sure childproperties is not null if there is children (for join resolver)
+                v.childProperties_ = new List<PhysicProperty>(new PhysicProperty[physic.children_?.Count ?? 0]);
+
+                // always update the null property min member
+                if (physic.Cost() < mincost)
                 {
                     mincost = physic.Cost();
-                    minMember_ = v;
+                    minMember_[new PhysicProperty()] = new Tuple<CGroupMember, double>(v, mincost);
+                }
+
+                // directly supply the property
+                if (required?.IsPropertySupplied(physic) ?? true && physic.Cost() < propmincost)
+                {
+                    propmincost = physic.Cost();
+                    minMember_[required] = new Tuple<CGroupMember, double>(v, propmincost);
+                }
+                // property is enforced
+                else if (required != null && physic.Cost() + required.OrderEnforcement(physic).Cost() < propmincost)
+                {
+                    propmincost = physic.Cost() + required.OrderEnforcement(physic).Cost();
+                    var enforced = new CGroupMember(required.OrderEnforcement(physic), this);
+                    minMember_[required] = new Tuple<CGroupMember, double>(enforced, propmincost);
                 }
             }
 
-            minIncCost_ = mincost;
+            // record the null property minimum inclusive cost
+            if(mincost > 0) minIncCost_ = mincost;
+
             Debug.Assert(IsSolverOptimizedGroup() ||
-                minMember_.physic_.InclusiveCost() == minIncCost_);
-            return minMember_;
+                minMember_[new PhysicProperty()].Item1.physic_.InclusiveCost() == minIncCost_);
+            return minMember_[required ?? new PhysicProperty()].Item1;
         }
 
-        public CGroupMember CalculateMinInclusiveCostMember()
+        internal void CalculateMemberCosts(PhysicProperty required, CGroupMember member,
```

Comment:
header comment

Reply:
let's try to pull up this logic with the parent (as it recursively called parent) and simplify with 3=>2 logics indicated in your checkin comment.